### PR TITLE
BookServiceTests

### PR DESCRIPTION
**Pull request by shumYevh**
зображення_2024-06-25_144044065
зображення_2024-06-25_144208557
зображення_2024-06-25_144401676
I didn`t test repositories cause i use default JPA queries (book, category)
Instead i implemented tests for orderItem repository, cause it have custom queries.

### CODE REVIEW

File: src/test/resources/database/book/add_SpringBoot_book_to_books_table.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+INSERT INTO books (title, author, isbn, price, description, cover_image, is_deleted)
+VALUES ('Learning Spring Boot', 'John Doe', '9781234567897', 29.99, 'A comprehensive guide to Spring Boot', 'url://spring_boot_cover.jpg', false);
```
Comment (Vasul-Pankiv) [inline src/test/resources/database/book/add_SpringBoot_book_to_books_table.sql]: each file has to end with an empty line

File: src/test/java/org/example/bookweb/controller/CategoryControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+                .setName("Fantasy")
+                .setDescription("Books about fantastic worlds"));
+        expected.add(new CategoryDto()
+                .setId(2L)
+                .setName("Horror")
+                .setDescription("Very scared books"));
+        expected.add(new CategoryDto()
+                .setId(3L)
+                .setName("Romantic")
+                .setDescription("Beautiful books with love"));
```
Comment (Vasul-Pankiv) [inline src/test/java/org/example/bookweb/controller/CategoryControllerTest.java]: create an utility class to provide tests with tests data

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

File: src/test/java/org/example/bookweb/controller/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    @SneakyThrows
+    private static void tearDown(DataSource dataSource) {
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(
+                    connection,
+                    new ClassPathResource("database/book/remove_all_books.sql")
+            );
+        }
+    }
```
Comment (MentorMateAcademy) [inline src/test/java/org/example/bookweb/controller/BookControllerTest.java]: Private method after public

File: src/test/java/org/example/bookweb/controller/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON)
+                )
+                .andExpect(status().isCreated())
+                .andReturn();
+
+        //Then
+        BookDto actual = objectMapper
+                .readValue(result.getResponse().getContentAsString(), BookDto.class);
+        EqualsBuilder.reflectionEquals(expected, actual, "id");
```
Comment (MentorMateAcademy) [inline src/test/java/org/example/bookweb/controller/BookControllerTest.java]:
```suggestion

        MvcResult result = mockMvc.perform(post("/books")
                        .content(jsonRequest)
                        .contentType(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().isCreated())
                .andReturn();

        //When
        BookDto actual = objectMapper
                .readValue(result.getResponse().getContentAsString(), BookDto.class);
        //Then
        EqualsBuilder.reflectionEquals(expected, actual, "id");
```

Comment (MentorMateAcademy) [reply src/test/java/org/example/bookweb/controller/BookControllerTest.java]: Fix all tests by this pattern

File: src/test/java/org/example/bookweb/controller/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    @Test
+    @DisplayName("Update existing book")
+    public void put_validRequestDto_Success() throws Exception {
+        long bookId = 99L;
+        CreateBookRequestDto requestDto = new CreateBookRequestDto()
+                .setTitle("NEW Test Title")
+                .setAuthor("NEW Test Author")
+                .setIsbn("0-7290-0264-0")
+                .setPrice(BigDecimal.valueOf(100))
+                .setCoverImage("url://NEW_TestCoverImage");
```
Comment (MentorMateAcademy) [inline src/test/java/org/example/bookweb/controller/BookControllerTest.java]: Save values to constants and reuse this

File: src/test/java/org/example/bookweb/controller/BookControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    }, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = {
+            "classpath:database/book/remove_testBook_from_books.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    @WithMockUser(username = "adminUser", authorities = {"ADMIN"})
+    @Test
+    @DisplayName("Search existing book")
+    public void search_validRequest_Success() throws Exception {
+        //Given
+        String testBookTitle = "Test Title";
```
Comment (MentorMateAcademy) [inline src/test/java/org/example/bookweb/controller/BookControllerTest.java:201]: only given you write, use same style

File: src/test/java/org/example/bookweb/service/BookServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+
+@ExtendWith(MockitoExtension.class)
+public class BookServiceTest {
+
```
Comment (MentorMateAcademy) [inline src/test/java/org/example/bookweb/service/BookServiceTest.java]:
```suggestion
```

File: src/test/java/org/example/bookweb/utils/TestDataUtil.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+import org.example.bookweb.dto.category.CreateCategoryRequestDto;
+
+public class TestDataUtil {
+    public static CreateBookRequestDto getTestCreateBookRequestDto() {
+        return new CreateBookRequestDto()
+                .setTitle("Test Title")
+                .setAuthor("Test Author")
+                .setIsbn("0-7290-0264-0")
+                .setPrice(BigDecimal.valueOf(50))
+                .setCoverImage("url://TestCoverImage");
```
Comment (MentorMateAcademy) [inline src/test/java/org/example/bookweb/utils/TestDataUtil.java]: Constants

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**